### PR TITLE
Tied params cleanup

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR is the first in a series that aims to clean up the content of the variables `_keys_to_ignore_xxx` and the logic of warnings sent to the user resulting from them. In particular unless the model implement a hack like RoBERTa, a trained version with no shared version might not save the decoder...

One use of these variables is to detect "normal" shared weights and have a way to remove all but one of those shared weights for safetensors serialization, but the variable used (`_keys_to_ignore_unexepected`) is very noisy (it contains names of weights that are not shared). This PR introduces a variable `_tied_weights_keys` which will be used for that purpose (and in other places of the cleanup later on). It contains the named of the weights that are tied to other weights and that are safe to dismiss when saving (we won't dismiss them on the torch.save side since `torch.save` won't duplicate memory but will use it for `safetensors`). This PR also introduces a test that checks the content of this variable is correct.

Due to the very large number of models having potential shared weights, this PR limits itself at the introduction of this variable, proper fill for all existing models and test. The rest will follow in subsequent PRs.

cc @Narsil for info

### CODE REVIEW

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -1662,6 +1664,34 @@ def test_can_use_safetensors(self):
...
 
+    def test_tied_weights_keys(self):
+        config, _ = self.model_tester.prepare_config_and_inputs_for_common()
+        config.tie_word_embeddings = True
+        for model_class in self.all_model_classes:
+            model_tied = model_class(config)
+
+            ptrs = collections.defaultdict(list)
+            for name, tensor in model_tied.state_dict().items():
+                ident = (tensor.data_ptr(), tensor.device, tensor.shape, tensor.stride())
```
Comment (Narsil) [inline tests/test_modeling_common.py:1675]: This will only catch  exactly shared tensors, never partial of transposed sharing.

We shouldn't have it in our code, but since this is a test, maybe making it more robust could be good (so we can catch whenever a new model with finer sharing comes along).

`(tensor.untyped_storage().data_ptr(), tensor.untyped_storage().nbytes())` is more robust. (There's no stride, nor shape on storage). (`untyped_storage` is 2.0 otherwise you have to use `.storage().bytes()` iirc.

This doesn't actually capture shared/non-shared storages faking (@thomasw21) but I think it's beyond the point.
There are ways to make 2 different storages actually point to the same memory (without necessary sharing either pointer nor size), leading to all sorts of weird things.

Comment (thomasw21) [reply tests/test_modeling_common.py:1675]: I guess this is fine here since the `_tied_weights_keys` mechanism shares not only the storage but also the entire `nn.Parameter` no?

Comment (thomasw21) [reply tests/test_modeling_common.py:1675]: Also I guess you could just use `id` no?

Comment (Narsil) [reply tests/test_modeling_common.py:1675]: >  So maybe let's factor that in a util function we reuse for this test, the safetensors check for duplicates and the checkpoint sharding if that's the strongest way?

Seems good to me. At least we have a unique source for pointer sharing.

Comment (thomasw21) [reply tests/test_modeling_common.py:1675]: > The simple id check is not sufficient @thomasw21 : it doesn't catch the tied weights that DeformableDetr has (I compared the output for all models).

Hum that seems very suspicious to me. I've had issues before where if you had 2 parameters (with different ids) that share the underlying tensors/storage ... then gradients where pretty weird ...

Comment (sgugger) [reply tests/test_modeling_common.py]: The corresponding model is used so little it wouldn't surprise me if training it was not working. That's [how](https://github.com/huggingface/transformers/blob/3723329d014a7b144863e597ea4fe6de5e6a8279/src/transformers/models/deformable_detr/modeling_deformable_detr.py#L308) it clones the weights, which is indeed pretty ugly

Comment (thomasw21) [reply tests/test_modeling_common.py]: Ah very good find!
```python
import copy

from torch import nn

def main():
    model = nn.Linear(3,5)
    new_model = copy.deepcopy(model)
    # Theses are different parameters ....
    assert id(model.weight) != id(new_model.weight)
    
if __name__ == "__main__":
    main()
```

**Review by LysandreJik — APPROVED** [review]
Ok this looks good to me! Looking forward to the next PR
